### ISSUE TITLE
[bug] ParameterHelper.cs Creates Duplicate Parameters and Causes ArgumentException

### ISSUE THREAD

**Issue by soelinn** [schotime/NPoco#374]
# Bug Description

**ArgumentException: The SqlParameter is already contained by another SqlParameterCollection.** is thrown when calling `db.FetchMultiple<T1,T2,T3>()` with the following SQL query with **NPoco 3.4.8**.

```cs
var strSql = @"
SELECT * FROM [dbo].[Table1] WHERE [Column1] = @Param1 AND [Column2] <= @Param2 
SELECT * FROM [dbo].[Table2] WHERE [Column1] = @Param1 AND [Column2] <= @Param2
SELECT * FROM [dbo].[Table3] WHERE [Column1] = @Param1 AND [Column2] <= @Param2";

var parameters = new {
    Param1 = new SqlParameter("@Param1", SqlDbType.VarChar, 20) { Value = "abcd" },
    Param2 = new SqlParameter("@Param1", SqlDbType.Date) { Value = DBNull.Value }
};

db.FetchMultiple<MyT1,MyT2,MyT3>(strSql, parameters);
```

# Source of Bug

[https://github.com/schotime/NPoco/blob/master/src/NPoco/ParameterHelper.cs](https://github.com/schotime/NPoco/blob/master/src/NPoco/ParameterHelper.cs) class's `ProcessParams` method creates **duplicate** parameters when called with the query above.

The `List<object> args_dest` gets the following items in it with these.

```cs
new SqlParameter("@Param1" ...),
new SqlParameter("@Param2" ...),
new SqlParameter("@Param1" ...),
new SqlParameter("@Param2" ...),
new SqlParameter("@Param1" ...),
new SqlParameter("@Param2" ...),
```
These `SqlParameter`s with the same name causes the `ArgumentException` when added to the `SqlCommand` later on.

# Test Code to Reproduce Issue

Copy and paste the following code into **LinqPad** to execute.
The following code is just a stripped-down version based on the current **master** branch so it can be easily executed to test without having to spend/waste time to create a VisualStudio project from scratch.

```cs
/*
// Reference these namespaces.

System
System.Collections
System.Collections.Generic
System.Linq
System.Reflection
System.Text
System.Text.RegularExpressions
*/

void Main()
{
	var _sql = @"
SELECT * FROM [dbo].[Table1] WHERE [Column1] = @Param1 AND [Column2] <= @Param2 
SELECT * FROM [dbo].[Table2] WHERE [Column1] = @Param1 AND [Column2] <= @Param2
SELECT * FROM [dbo].[Table3] WHERE [Column1] = @Param1 AND [Column2] <= @Param2";

	var args_src = new object[] {
		new {
			Param1 = new SqlParameter("@Param1", SqlDbType.VarChar, 20) { Value = "abcd" },
			Param2 = new SqlParameter("@Param2", SqlDbType.Date) { Value = DBNull.Value }
		}
	};
	var args_dest = new List<object>();

	var output = ProcessParams(_sql, args_src, args_dest);

	// Use LinqPad to see (or) do Console.WriteLine
	output.Dump();
	args_dest.Dump();
}

#region ParameterHelper.cs ProcessParams
public static string ProcessParams(string _sql, object[] args_src, List<object> args_dest)
{
	var regex = new Regex(@"(?<!@)@\w+", RegexOptions.Compiled);

	return regex.Replace(_sql, m =>
	{
		string param = m.Value.Substring(1);

		object arg_val;

		int paramIndex;
		if (Int32.TryParse(param, out paramIndex))
		{
			// Numbered parameter
			if (paramIndex < 0 || paramIndex >= args_src.Length)
				throw new ArgumentOutOfRangeException(String.Format("Parameter '@{0}' specified but only {1} parameters supplied (in `{2}`)", paramIndex, args_src.Length, _sql));
			arg_val = args_src[paramIndex];
		}
		else
		{
			// Look for a property on one of the arguments with this name
			bool found = false;
			arg_val = null;
			foreach (var o in args_src)
			{
				var dict = o as IDictionary;
				if (dict != null)
				{
					Type[] arguments = dict.GetType().GetGenericArguments();

					if (arguments[0] == typeof(string))
					{
						var val = dict[param];
						if (val != null)
						{
							found = true;
							arg_val = val;
							break;
						}
					}
				}

				var pi = o.GetType().GetProperty(param);
				if (pi != null)
				{
					arg_val = pi.GetValue(o, null);
					found = true;
					break;
				}
			}

			if (!found)
				throw new ArgumentException(String.Format("Parameter '@{0}' specified but none of the passed arguments have a property with this name (in '{1}')", param, _sql));
		}

		// Expand collections to parameter lists
		if ((arg_val as System.Collections.IEnumerable) != null &&
			(arg_val as string) == null &&
			(arg_val as byte[]) == null)
		{
			var sb = new StringBuilder();
			foreach (var i in arg_val as System.Collections.IEnumerable)
			{
				var indexOfExistingValue = args_dest.IndexOf(i);
				if (indexOfExistingValue >= 0)
				{
					sb.Append((sb.Length == 0 ? "@" : ",@") + indexOfExistingValue);
				}
				else
				{
					sb.Append((sb.Length == 0 ? "@" : ",@") + args_dest.Count);
					args_dest.Add(i);
				}
			}
			if (sb.Length == 0)
			{
				Type type = typeof(string);
				var t = arg_val.GetType();
				if (t.IsArray)
					type = t.GetElementType();
				else if (t.IsOrHasGenericInterfaceTypeOf(typeof(IEnumerable<>)))
					//else if (t.getType())
					type = t.GetGenericArguments().First();

				sb.AppendFormat($"select @{args_dest.Count} /*poco_dual*/ where 1 = 0");
				args_dest.Add(MappingHelper.GetDefault(type));
			}
			return sb.ToString();
		}
		else
		{
			args_dest.Add(arg_val);
			return "@" + (args_dest.Count - 1).ToString();
		}
	});
}
#endregion

#region Reflectionhelper
public static class ReflectionUtils
{
	public static List<MemberInfo> GetFieldsAndProperties<T>(BindingFlags bindingAttr)
	{
		return GetFieldsAndProperties(typeof(T), bindingAttr);
	}

	public static List<MemberInfo> GetFieldsAndPropertiesForClasses(Type type)
	{
		if (type.GetTypeInfo().IsValueType || type == typeof(string) || type == typeof(byte[]) || type == typeof(Dictionary<string, object>) || type.IsArray)
			return new List<MemberInfo>();

		return GetFieldsAndProperties(type);
	}

	public static List<MemberInfo> GetFieldsAndProperties(Type type)
	{
		return GetFieldsAndProperties(type, BindingFlags.Instance | BindingFlags.Public);
	}

	public static List<MemberInfo> GetFieldsAndProperties(Type type, BindingFlags bindingAttr)
	{
		List<MemberInfo> targetMembers = new List<MemberInfo>();

		targetMembers.AddRange(type.GetFields(bindingAttr).Where(x => !x.IsInitOnly).ToArray());
		targetMembers.AddRange(type.GetProperties(bindingAttr));

		return targetMembers;
	}

	public static Type GetMemberInfoType(this MemberInfo member)
	{
		Type type;
		if (member is FieldInfo)
			type = ((FieldInfo)member).FieldType;
		else if (member is PropertyInfo)
			type = ((PropertyInfo)member).PropertyType;
		else if (member == null)
			type = typeof(object);
		else
			throw new NotSupportedException();

		return type;
	}

	public static bool IsField(this MemberInfo member)
	{
		return member is FieldInfo;
	}

	public static MethodInfo GetSetMethodOnDeclaringType(this PropertyInfo propertyInfo)
	{
		var methodInfo = propertyInfo.GetSetMethod(true);
		return methodInfo ?? propertyInfo
								.DeclaringType
								.GetProperty(propertyInfo.Name)
								.GetSetMethod(true);
	}

	public static bool IsOrHasGenericInterfaceTypeOf(this Type type, Type genericTypeDefinition)
	{
		return type.GetTypeWithGenericTypeDefinitionOf(genericTypeDefinition) != null;
	}

	public static Type GetTypeWithGenericTypeDefinitionOf(this Type type, Type genericTypeDefinition)
	{
		foreach (var t in type.GetInterfaces())
		{
			if (t.GetTypeInfo().IsGenericType && t.GetGenericTypeDefinition() == genericTypeDefinition)
			{
				return t;
			}
		}

		var genericType = type.GetGenericType();
		if (genericType != null && genericType.GetGenericTypeDefinition() == genericTypeDefinition)
		{
			return genericType;
		}

		return null;
	}

	public static Type GetGenericType(this Type type)
	{
		while (type != null)
		{
			if (type.GetTypeInfo().IsGenericType)
				return type;

			type = type.GetTypeInfo().BaseType;
		}
		return null;
	}

	public static Type GetTypeWithInterfaceOf(this Type type, Type interfaceType)
	{
		if (type == interfaceType) return interfaceType;

		foreach (var t in type.GetInterfaces())
		{
			if (t == interfaceType)
				return t;
		}

		return null;
	}

	public static bool IsOfGenericType(this Type instanceType, Type genericType)
	{
		Type type = instanceType;
		while (type != null)
		{
			if (type.GetTypeInfo().IsGenericType &&
				type.GetGenericTypeDefinition() == genericType)
			{
				return true;
			}
			type = type.GetTypeInfo().BaseType;
		}

		foreach (var i in instanceType.GetInterfaces())
		{
			if (i.GetTypeInfo().IsGenericType && i.GetGenericTypeDefinition() == genericType)
			{
				return true;
			}
		}
		return false;
	}
}
#endregion

#region MappingHelper
// Define other methods and classes here
public class MappingHelper
{
	public static object GetDefault(Type type)
	{
		if (type.GetTypeInfo().IsValueType)
		{
			return Activator.CreateInstance(type);
		}
		return null;
	}
}
#endregion
```

# DNVM Issue for Visual Studio

When I open the **NPoco.sln** in my Visual Studio 2015, I get these errors for the **DNVM**.
I would have patched the code and created a Pull Request if I could open the project without any issue.

I didn't feel like wasting time to track down and install the missing obsolete .NET Core setup files to be able to compile the solution.

Anyhow, I suggest it is time to upgrade/fix the project to the latest .NET Core project structure (i.e., the MSBuild-based one) since VS2017 is going to be released in about 2 weeks from now. That way, more people can help out and contribute to this project.

```sh
Installing to C:\Users\me\.dnx\runtimes\dnx-clr-win-x86.1.0.0-beta8
Adding C:\Users\me\.dnx\runtimes\dnx-clr-win-x86.1.0.0-beta8\bin to process PATH
Installing to C:\Users\me\.dnx\runtimes\dnx-coreclr-win-x86.1.0.0-beta8
Adding C:\Users\me\.dnx\runtimes\dnx-coreclr-win-x86.1.0.0-beta8\bin to process PATH
Compiling native images for dnx-coreclr-win-x86.1.0.0-beta8 to improve startup performance...
Downloading dnx-clr-win-x86.1.0.0-preview2-003121 from https://www.nuget.org/api/v2
The server returned a 404 (NotFound). This is most likely caused by the feed not having the version that you typed. Check that you typed the right version and try again. Other possible causes are the feed doesn't have a DNX of the right name format or some other error caused a 404 on the server.
At C:\Program Files\Microsoft DNX\Dnvm\dnvm.ps1:558 char:13
+             throw "The server returned a 404 (NotFound). This is most ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (The server retu... on the server.:String) [], RuntimeException
    + FullyQualifiedErrorId : The server returned a 404 (NotFound). This is most likely caused by the feed not having the version that you typed. Check that you typed the right version and try again. Other possible causes are the feed doesn't have a DNX of the right name format or some other error caused a 404 on the server.
 
Downloading dnx-clr-win-x86.1.0.0-preview2-003121 from https://www.myget.org/F/aspnetvnext/api/v2
The server returned a 404 (NotFound). This is most likely caused by the feed not having the version that you typed. Check that you typed the right version and try again. Other possible causes are the feed doesn't have a DNX of the right name format or some other error caused a 404 on the server.
At C:\Program Files\Microsoft DNX\Dnvm\dnvm.ps1:558 char:13
+             throw "The server returned a 404 (NotFound). This is most ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (The server retu... on the server.:String) [], RuntimeException
    + FullyQualifiedErrorId : The server returned a 404 (NotFound). This is most likely caused by the feed not having the version that you typed. Check that you typed the right version and try again. Other possible causes are the feed doesn't have a DNX of the right name format or some other error caused a 404 on the server.
```

Please let me know if you have any question.

**Comment by schotime** [comment]
Firstly... this is a bug, and I have a fix that I will release in a version today.
Also, It doesn't use dnx to build. It uses the RTM of .NET Core.

**Comment by schotime** [comment]
Released in 3.5.0
